steering_controllers/src/steering_controllers.cpp

```diff
@@ -0,0 +1,489 @@
+    covariance_controller[diagonal_index] = params_.twist_covariance_diagonal[index];
+  }
+  RCLCPP_INFO(get_node()->get_logger(), "configure successful");
+  return controller_interface::CallbackReturn::SUCCESS;
+}
+
+void SteeringControllers::reference_callback(const std::shared_ptr<ControllerReferenceMsg> msg)
+{
+  // if no timestamp provided use current time for command timestamp
+  if (msg->header.stamp.sec == 0 && msg->header.stamp.nanosec == 0u)
+  {
+    RCLCPP_WARN(
+      get_node()->get_logger(),
+      "Timestamp in header is missing, using current time as command timestamp.");
+    msg->header.stamp = get_node()->now();
+  }
+  const auto age_of_last_command = get_node()->now() - msg->header.stamp;
+
+  if (ref_timeout_ == rclcpp::Duration::from_seconds(0) || age_of_last_command <= ref_timeout_)
+  {
+    input_ref_.writeFromNonRT(msg);
+  }
+  else
+  {
+    RCLCPP_ERROR(
+      get_node()->get_logger(),
+      "Received message has timestamp %.10f older for %.10f which is more then allowed timeout "
+      "(%.4f).",
+      rclcpp::Time(msg->header.stamp).seconds(), age_of_last_command.seconds(),
+      ref_timeout_.seconds());
+  }
+}
+
+void SteeringControllers::reference_callback_unstamped(
+  const std::shared_ptr<geometry_msgs::msg::Twist> msg)
+{
+  auto twist_stamped = *(input_ref_.readFromNonRT());
+  twist_stamped->header.stamp = get_node()->now();
+  // if no timestamp provided use current time for command timestamp
+  if (twist_stamped->header.stamp.sec == 0 && twist_stamped->header.stamp.nanosec == 0u)
+  {
+    RCLCPP_WARN(
+      get_node()->get_logger(),
+      "Timestamp in header is missing, using current time as command timestamp.");
+    twist_stamped->header.stamp = get_node()->now();
+  }
+
+  const auto age_of_last_command = get_node()->now() - twist_stamped->header.stamp;
+
+  if (ref_timeout_ == rclcpp::Duration::from_seconds(0) || age_of_last_command <= ref_timeout_)
+  {
+    twist_stamped->twist = *msg;
+  }
+  else
+  {
+    RCLCPP_ERROR(
+      get_node()->get_logger(),
+      "Received message has timestamp %.10f older for %.10f which is more then allowed timeout "
+      "(%.4f).",
+      rclcpp::Time(twist_stamped->header.stamp).seconds(), age_of_last_command.seconds(),
+      ref_timeout_.seconds());
+  }
+}
+
+controller_interface::InterfaceConfiguration SteeringControllers::command_interface_configuration()
+  const
+{
+  controller_interface::InterfaceConfiguration command_interfaces_config;
+  command_interfaces_config.type = controller_interface::interface_configuration_type::INDIVIDUAL;
+  command_interfaces_config.names.reserve(nr_cmd_itfs_);
+
+  for (size_t i = 0; i < params_.rear_wheel_names.size(); i++)
+  {
+    command_interfaces_config.names.push_back(
+      params_.rear_wheel_names[i] + "/" + hardware_interface::HW_IF_VELOCITY);
+  }
+  for (size_t i = 0; i < params_.front_steer_names.size(); i++)
```

Comment:
Since we are also planning to enable steering on the rear axis. Would it make sense to rename this parameter, for example: 

```suggestion
  for (size_t i = 0; i < params_.front_wheel_names.size(); i++)
```

Reply:
This is being updated as HW_IF_POSITION at the moment. That's why it's more important to make a distinction between the joint's purpose rather than the location in a configuration. We should probably rename it as steer_names and traction_names, what do you think about this?